functional/connectivity-loss_test.go

```diff
@@ -0,0 +1,229 @@
+		}
+	}
+
+	// Cut connection to etcd.
+	//
+	// We use REJECT here, so fleet knows immediately that it's disconnected, rather than waiting for a timeout.
+	if _, err = cluster.MemberCommand(m0, "sudo", "iptables", "-I", "OUTPUT", "-p", "tcp", "-m", "multiport", "--dports=2379,4001", "-j", "REJECT"); err != nil {
+		t.Fatal(err)
+	}
+
+	// Wait long enough to be reasonably confident that no more state changes will happen.
+	ttl, _ := time.ParseDuration(util.FleetTTL)
+	agentReconcileInterval := 5 * time.Second
+	slack := 2 * time.Second
+
+	time.Sleep(ttl + agentReconcileInterval + slack)
+
+	// Check unit state after connection loss.
+	//
+	// Note: we cannot use fleetctl to check the state here,
+	// as fleet is not available to give us this information...
+	// We have to go deeper, and try to obtain the information from systemd directly.
+	stdout, err := cluster.MemberCommand(m0, "systemctl", "list-unit-files", "-t", "service", "--no-legend", "single@*.service", "global@*.service")
+	stdout = strings.TrimSpace(stdout)
+	if err != nil {
+		t.Fatalf("Failed to retrieve systemd unit file states: %v", err)
+	}
+	actualSystemdFiles := map[string]string{}
+	if stdout != "" {
+		for _, line := range strings.Split(stdout, "\n") {
+			fields := strings.Fields(line)
+			actualSystemdFiles[fields[0]] = fields[1]
+		}
+	}
+	if !reflect.DeepEqual(actualSystemdFiles, expectedSystemdFiles) {
+		t.Fatalf("Units files not in expected state after loosing connectivity.\nExpected: %v\nActual: %v", expectedSystemdFiles, actualSystemdFiles)
+	}
+
+	stdout, err = cluster.MemberCommand(m0, "systemctl", "list-units", "-t", "service", "--no-legend", "single@*.service", "global@*.service")
+	stdout = strings.TrimSpace(stdout)
+	if err != nil {
+		t.Fatalf("Failed to retrieve systemd unit states: %v", err)
+	}
+	actualSystemdStates := map[string][]string{}
+	if stdout != "" {
+		for _, line := range strings.Split(stdout, "\n") {
+			fields := strings.Fields(line)
+			actualSystemdStates[fields[0]] = fields[1:4]
+		}
+	}
+	if !reflect.DeepEqual(actualSystemdStates, expectedSystemdStates) {
+		t.Fatalf("Units not in expected state after loosing connectivity.\nExpected: %v\nActual: %v", expectedSystemdStates, actualSystemdStates)
+	}
+
+	// Restore etcd connection.
+	if _, err = cluster.MemberCommand(m0, "sudo", "iptables", "-D", "OUTPUT", "-p", "tcp", "-m", "multiport", "--dports=2379,4001", "-j", "REJECT"); err != nil {
+		t.Fatal(err)
+	}
+
+	// Again, wait long enough to be reasonably confident that no more state changes will happen.
+	//
+	// Here this should cover the time for fleet to realise connectivity is back,
+	// and for the Agent to complete the second run after reconnection.
+	//
+	// (Unlike for the first run immediately after connectivity is back, by the time of the second run,
+	// Engine leadership and Engine reconciliation should have been sorted out,
+	// and everything should be back to normal...)
+	time.Sleep(ttl + agentReconcileInterval + slack)
```

Comment:
The entire test takes about 25s, including two waits of 10s each.

It's a bit longer than I would like, but not terribly so (many other tests take > 10s); and I can't think of a way to speed it up without forsaking reliability...